### PR TITLE
Remove use of data() in optimizers

### PR DESCRIPTION
**Pull request by goldsborough**
After talking to users of the C++ API we found that having the tensor type be `autograd::Variable` causes more complications than having it be `at::Tensor`. It used to be a problem because `at::Tensor` didn't have the "autograd API" of variable (e.g. `detach()` or `grad()` methods), but those methods are now on `at::Tensor`. As such, we want to make a last big breaking change to have the tensor type be `at::Tensor`, while factory methods like `torch::ones` will return `Variable`s disguised as `at::Tensor`. This will make many things easier, like calling functions in ATen that take vectors of tensors.

This PR makes a small step in this direction by updating the optimizer classes to not use `.data()` on `Variable` to access the underlying `at::Tensor`. Using `.data()` is effectively a hack to work around our modification rules for tensors that require grad. The proper way of doing things is to use `with torch.no_grad` or equivalently `NoGradGuard` in C++ to guard in-place operations.

The next step can then simply redefine `torch::Tensor` to be `at::Tensor`. This transition should be smooth, since all methods available on `Variable` are at this point available on `at::Tensor`.

For this PR I:

1. Modified the implementations of optimizers to not use `.data()`. This means the implementations are now different from PyTorch, which still uses the legacy method of using `.data`.
2. To properly verify (1), I added more fine-grained test cases to our optimizer tests, e.g. `SGD` with and without `weight_decay`, then with `nesterov` etc. Generally more tests = more happy!
3. Minor cleanup of the optimizer codebase

@ebetica @apaszke

### CODE REVIEW

File: torch/csrc/api/src/optim/optimizer.cpp [hunk 1]
```diff
@@ -29,10 +29,9 @@ void OptimizerBase::add_parameters(const ParameterCursor& cursor) {
 
 void OptimizerBase::zero_grad() {
   for (auto& parameter : parameters_) {
-    auto& grad = parameter.grad();
-    if (grad.defined()) {
-      grad = grad.detach();
-      Tensor(grad).data().zero_();
+    if (parameter.grad().defined()) {
+      parameter.grad().detach_();
+      parameter.grad().zero_();
```
Comment (ebetica) [inline torch/csrc/api/src/optim/optimizer.cpp]: I wonder if this should also be `detach().zero()`? Or if not, could you add a test like:

```
auto model = Something();
auto optim = SGD(model->parameters(), 1e-3);
optim->zero_grad();
auto random_operation = model->parameters()->begin()->value.grad() * 2;
```

My hypothesis is that the grad() will point the one before the zero_, and then the code that detects if an in-place op was done will run and it will runtime_error.

File: test/cpp/api/optim_baseline.h [hunk 2]
```diff
@@ -7,75 +7,351 @@
 namespace expected_parameters {
 
 static std::vector<std::vector<torch::Tensor>> Adam = {
+  {
```
Comment (ebetica) [inline test/cpp/api/optim_baseline.h]: Might be valuable to comment how you generated these?

File: torch/csrc/api/include/torch/serialization.h [hunk 3]
```diff
@@ -237,4 +266,51 @@ void load(Archive& archive, torch::Tensor& tensor) {
...
+  tensor.resize_(sizes);
+
+  if (tensor.type().is_cuda()) {
+    // should actually use cudamemcpy probably
+    auto cputensor = torch::empty(sizes, tensor.dtype());
+    agimpl::loadBinary(
+        archive,
+        cputensor.data_ptr(),
+        cputensor.numel() * cputensor.type().elementSizeInBytes());
+    tensor.copy_(cputensor.data());
```
Comment (ebetica) [inline torch/csrc/api/include/torch/serialization.h]: again, `detach().copy_()`?

File: torch/csrc/api/src/nn/modules/embedding.cpp [hunk 4]
```diff
@@ -20,7 +20,7 @@ EmbeddingImpl::EmbeddingImpl(EmbeddingOptions options)
 void EmbeddingImpl::reset() {
   weight = register_parameter(
       "weight", torch::empty({options.count_, options.dimension_}));
-  weight.data().normal_(0, 1);
+  weight.normal_(0, 1);
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply torch/csrc/api/src/nn/modules/embedding.cpp]: I put a `NoGradGuard` around it now

[the remark this answers is not in the document]
Comment (ebetica) [reply torch/csrc/api/src/nn/modules/embedding.cpp]: I think @zdevito said this doesn't prevent the not-a-leaf checks? Or I'm confused about when this happens and when it doesn't.

File: test/cpp/api/optim_baseline.h (hunk 2 above)
Comment (goldsborough) [reply test/cpp/api/optim_baseline.h]: There's `// @generated from test/cpp/api/optim_baseline.py` at the top :)

Comment (ebetica) [reply test/cpp/api/optim_baseline.h]: Oops, thanks!

File: torch/csrc/api/src/nn/modules/embedding.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ebetica) [reply torch/csrc/api/src/nn/modules/embedding.cpp]: Oh, I see, so if you have a NoGradGuard, the graph is never recorded, so you don't have this issue. This means `NoGradGuard g; weight.zero_()` and `weight.detach().zero_()` are equivalent, except that the second incantation creates a graph node that's immediately destroyed since the result is never assigned to anything.

File: torch/csrc/api/src/optim/optimizer.cpp (hunk 1 above)
Comment (ebetica) [reply torch/csrc/api/src/optim/optimizer.cpp]: Alternatively, have a NoGradGuard here as well.

Comment (ebetica) [reply torch/csrc/api/src/optim/optimizer.cpp]: I thought double backwards was possible? In which case the grad might be a node in an autograd graph, no?

**Review by ebetica — APPROVED** [review]
